src/core/numeric_text_file.cpp

```diff
@@ -255,8 +280,11 @@ void NumericTextFile::WriteLine(double* data_array) {
     }
 }
 
+template void NumericTextFile::WriteLine<float>(float* data_array);
+template void NumericTextFile::WriteLine<double>(double* data_array);
```

Comment:
Not quite.

This just forces the compiler to instantiate this function for these types. Since we compile a library first, the compiler won't see all cases where these functions are called. 

Giving a string type in a separate compilation unit *would* compile, but would fail at link time (because there are only float/double binaries to link to.)

Using a different approach you can limit what templates are ever instantiated. In cpp20 (which we don't have) there is a feature called "concepts" to handle this. They also make much easier to under compiler errors.

Before cpp20 you can use a work around based on the "SFINAE" principle. ie "Substitution Failure is Not An Error."

From wiki 

"If an error occurs during the substitution of a set of arguments for any given template, the compiler removes the potential overload from the candidate set instead of stopping with a compilation error, provided the substitution error is one the C++ standard grants such treatment."

One way to implement this is via std::enable_if<> which can be setup to create a failed substitution for specific types. (The how there gets a bit complicated to type out on my phone)

But in a nutshell, you might have a templated helper function that returns void when enable if is true, but is undefined when false. You can then place that helper ad a template argument or return type (usually a trailing return) in the template you want to limit. So if you say

```cpp
template< T, std::enable_if<std::is_arithmetic<T>>::type = true> your_function( ) {}

```

Notes:

* this isn't the best way to do it 
* Rather than enable_if_t i used the explicit enable_if<>::type to point out how it works. If T is not arithmetic, then enable_if has no member called "type" and so the substitution fails and the compiler goes on its way

If you want to know more, I could point you to some better examples in our code base